lib/saluki-core/src/cache/mod.rs

```diff
@@ -0,0 +1,902 @@
+        // otherwise they would have already been merged when they were reclaimed.
+        let mut found_current = false;
+        let mut maybe_prev_entry = None;
+        let mut maybe_next_entry = None;
+
+        for entry in self.reclaimed.iter() {
+            match (found_current, entry == &curr_entry) {
+                // We haven't found our current entry yet, and this one directly precedes it.
+                (false, false) => {
+                    if entry.followed_by(&curr_entry) {
+                        maybe_prev_entry = Some(*entry);
+                    }
+                }
+                // We found our current entry, so we know to stop iterating after one more entry.
+                (false, true) => found_current = true,
+                // We found our current entry, and this one directly follows it.
+                (true, false) => {
+                    if curr_entry.followed_by(entry) {
+                        maybe_next_entry = Some(*entry);
+                        break;
+                    }
+                }
+                // Can't have found our current entry and then find it again.
+                (true, true) => unreachable!(),
+            }
+        }
+
+        // If we had no adjacent previous or next entries, then we're done.
+        if maybe_prev_entry.is_none() && maybe_next_entry.is_none() {
+            return;
+        }
+
+        self.reclaimed.remove(&curr_entry);
+
+        let (new_start, new_aligned) = match maybe_prev_entry {
+            None => (curr_entry.start, curr_entry.aligned),
+            Some(prev_entry) => {
+                self.reclaimed.remove(&prev_entry);
+                (prev_entry.start, prev_entry.aligned)
+            }
+        };
+
+        let new_end = match maybe_next_entry {
+            None => curr_entry.end,
+            Some(next_entry) => {
+                self.reclaimed.remove(&next_entry);
+                next_entry.end
+            }
+        };
+
+        self.reclaimed
+            .insert(ReclaimedEntry::new(new_start, new_end, new_aligned));
+    }
+
+    fn mark_for_reclamation(&mut self, header_ptr: NonNull<EntryHeader>) {
+        // See if the reference count is zero.
+        //
+        // Only interned string values (the frontend handle that wraps the pointer to a specific entry) can decrement
+        // the reference count for their specific entry when dropped, and only `InternerState` -- with its access mediated through a
+        // mutex -- can increment the reference count for entries. This means that if the reference count is zero, then
+        // we know that nobody else is holding a reference to this entry, and no concurrent call to `try_intern` could
+        // be updating the reference count, either... so it's safe to be marked as reclaimed.
+        //
+        // SAFETY: We know the pointer is well-aligned for `EntryHeader` and is initialized.
```

Comment:
You're not wrong. Running this through `loom` would be nice, and it might not be that hard to whip up a simple case to try and do just that....

I'm also open to ideas on how to shrink the context space to make it simpler. Like, one of the big issues that many of the safety aspects are built on knowing we never access `InternerState` except behind a lock, but you can't easily see that just by looking at `InternerState` in isolation.

Reply:
Ended up adding a Loom test in 6a7447b for this.